### PR TITLE
Support T5 Beam Search with DecoderMaskedMHA

### PR DESCRIPTION
**Pull request by wangyems**
### Description

tldr:
Latency improvement
t5-small: 37.8% 
t5-base: 24.5%

Benchmark on V100

Before:
T5-small
ORT {'test_times': 1, 'latency_variance': '0.00', 'latency_90_percentile': '104.74', 'latency_95_percentile': '104.74', 'latency_99_percentile': '104.74', 'average_latency_ms': '104.74', 'QPS': '19.10', 'parity': True}
T5-base
ORT {'test_times': 1, 'latency_variance': '0.00', 'latency_90_percentile': '200.93', 'latency_95_percentile': '200.93', 'latency_99_percentile': '200.93', 'average_latency_ms': '200.93', 'QPS': '9.95', 'parity': True}

After:
T5-small
ORT {'test_times': 1, 'latency_variance': '0.00', 'latency_90_percentile': '76.01', 'latency_95_percentile': '76.01', 'latency_99_percentile': '76.01', 'average_latency_ms': '76.01', 'QPS': '26.31', 'parity': True}
T5-base
ORT {'test_times': 1, 'latency_variance': '0.00', 'latency_90_percentile': '161.40', 'latency_95_percentile': '161.40', 'latency_99_percentile': '161.40', 'average_latency_ms': '161.40', 'QPS': '12.39', 'parity': True}

### CODE REVIEW

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 1]
```diff
@@ -250,13 +250,13 @@ def parse_arguments(argv: Optional[List[str]] = None) -> argparse.Namespace:
     model_group.set_defaults(past_present_share_buffer=False)
 
     model_group.add_argument(
-        "--use_decoder_masked_self_attention",
+        "--use_decoder_masked_attention",
         required=False,
         action="store_true",
         help="Uses `DecoderMaskedSelfAttention` to optimize the unidirectional decoding Attention computation. "
```
Comment (hariharans29) [inline onnxruntime/python/tools/transformers/convert_generation.py:256]: Help message needs update ?

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h [hunk 2]
```diff
@@ -214,6 +230,10 @@ Status BeamSearchT5<T>::Execute(const FeedsFetchesManager& encoder_feeds_fetches
   int iteration_counter = 0;
   std::vector<OrtValue> decoder_feeds;
   int current_length = parameters->sequence_length;
+
+  // TODO(tianleiwu): allocate fetches. use ping-pong buffers for past state.
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h:234]: I think there are some shared buffer for past in GPT-2. Does T5 have similar optimization?

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h [hunk 3]
```diff
@@ -251,6 +298,13 @@ Status BeamSearchT5<T>::Execute(const FeedsFetchesManager& encoder_feeds_fetches
       dumper->Print("decoder_feeds", i, true);
       dumper->Print("", decoder_feeds[i]);
     }
+    auto offset = decoder_subgraph_.GetFirstPastInputIndex() + 4 * decoder_subgraph_.num_layers;
+    dumper->Print("past_sequence_length", offset, true);
+    dumper->Print("", decoder_feeds[offset]);
+    dumper->Print("beam_width", offset+1, true);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h:304]: format code

File: onnxruntime/python/tools/transformers/convert_generation.py (hunk 1 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/convert_generation.py]: good catch

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h (hunk 2 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h:234]: yes, https://github.com/microsoft/onnxruntime/pull/15386/files#diff-ba924327241e2d12b13879b683af88533444b675fac501a60d42bc4528fef6c2R265-R277

File: onnxruntime/contrib_ops/cuda/decoder/fastertransformer_decoder_attention/decoder_masked_multihead_attention_impl.cu [hunk 4]
```diff
@@ -304,7 +304,7 @@ __global__ void masked_multihead_attention_kernel(DecoderMaskedMultiHeadAttentio
   int ti_end = ((tlength + K_PER_WARP - 1) / K_PER_WARP) * K_PER_WARP;
 
   // Iterate over the keys/timesteps to compute the various (Q*K^T)_{ti} values.
-  bool has_beams = params.cache_indir != nullptr;
+  bool has_beams = params.cache_indir != nullptr && !params.is_cross_attention;
```
Comment (hariharans29) [inline onnxruntime/contrib_ops/cuda/decoder/fastertransformer_decoder_attention/decoder_masked_multihead_attention_impl.cu:307]: Don't we support beam search with cross attention ? If so, how does it work with encoder-decoder architectures like T5 which by definiton have cross attention ?

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 5]
```diff
@@ -759,15 +759,15 @@ __global__ void UpdateDecoderMaskedSelfAttentionCacheIndirectionKernel(int32_t*
   const int batch_id = bb_id / beam_width;
   const int beam_id = bb_id % beam_width;
 
-  if (bb_id >= beam_width * batch_size || time_step >= current_length) {
+  if (bb_id >= beam_width * batch_size || time_step >= max_seq_length) {
```
Comment (hariharans29) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:762]: Can you please explain the changes here ? Why do we care about indirection indices for tokens after current_length ?

File: onnxruntime/contrib_ops/cuda/decoder/fastertransformer_decoder_attention/decoder_masked_multihead_attention_impl.cu (hunk 4 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/decoder/fastertransformer_decoder_attention/decoder_masked_multihead_attention_impl.cu:307]: The cross attention takes in static key/value cache and they are not involved in beam-indices based buffer switch.

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 5 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:762]: it's just for debug purposes (to not see uninitialized values)

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h:234]: Then we can remove the TODO comment.

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu]: That's true. Ideally, we would be memsetting the indirection buffer to all zeros before starting. That way, we can even avoid the first `if` where we assign the beam index to 0 for sequence indices < input sequence length. This is how FT does it. But, in any case, having un-initialized values in the indices >= current length should not matter as we shouldn't be using them anywhere.

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.h [hunk 6]
```diff
@@ -37,7 +38,9 @@ class T5DecoderSubgraph : public Subgraph {
       Stream* stream,
       bool use_sequence_as_input_ids,
       int cur_len,
-      transformers::Sequences& sequences);
+      transformers::Sequences& sequences,
+      int past_present_share_buffer_max_seq_len = -1,
+      bool add_beam_search_specific_inputs_for_decoder_masked_self_attention = false);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.h:43]: >add_beam_search_specific_inputs_for_decoder_masked_self_attention [](http://example.com/codeflow?start=11&length=65)

nit: the name is too long :)

Comment (yufenglee) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.h:43]: like, need_cache_dir?

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 7]
```diff
@@ -195,37 +220,73 @@ Status T5DecoderSubgraph::CreateInitialFeeds(
...
                                                      expanded_cache,
-                                                     false));
+                                                     false,
+                                                     use_max_seq_len ? past_present_share_buffer_max_seq_len : 0));
       }
       decoder_feeds.push_back(expanded_cache);
     }
   }
 
+  // TODO: This part shares the similar logic with CreateInitialFeeds() in subgraph_gpt.cc. We should refactor it.
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:261]: > // TODO: [](http://example.com/codeflow?start=1&length=9)

The function CreateInitialFeeds is a little bit too long. It would great if we can refactor the code instead of adding a TODO here, like moving this logic as a helper function in Subgraph

File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.h [hunk 8]
```diff
@@ -159,11 +159,15 @@ using UpdateDecoderFeedsFunc = std::function<Status(
...
     gsl::span<const int32_t> beam_indices,
+    gsl::span<const int32_t> beam_indices_gpu,
     int num_beams,
     int t5_decoder_first_past_input_idx,
     int t5_decoder_first_present_output_idx,
     bool use_sequence_as_input_ids,
     int current_length,
+    int input_sequence_len,
+    bool past_present_share_buffer,
+    bool has_beam_search_specific_inputs_for_decoder_masked_multihead_attention,
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.h:170]: >has_beam_search_specific_inputs_for_decoder_masked_multihead_attention [](http://example.com/codeflow?start=9&length=70)

nit: similar here. Name is too long

File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.h [hunk 9]
```diff
@@ -298,11 +303,15 @@ Status UpdateDecoderFeeds(
...
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.h:314]: >has_beam_search_specific_inputs_for_decoder_masked_multihead_attention [](http://example.com/codeflow?start=9&length=70)

ditto

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc (hunk 7 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:261]: sure

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.h (hunk 6 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.h:43]: sounds good

File: onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc [hunk 10]
```diff
@@ -47,6 +47,14 @@ namespace onnxruntime {
 namespace contrib {
 namespace GenerationCudaDeviceHelper {
 
+// This function assumes the Attention type is same as graph's past state type.
+// e.g In the case of past(fp32) -> cast to fp16 -> Attention(fp16), the reorder
+// function will use the fp32 chunk size and cause the model silently generates
+// the incorrect results.
+// TODO: Fix this issue. Either retrive the Attention op type from the graph or
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:54]: >TODO [](http://example.com/codeflow?start=3&length=4)

We need to experiment how the decoder_mask_attention kernel perform if we don't introduce this reorder.

**Review by yufenglee — APPROVED** [review]
:shipit:
